servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java

```diff
@@ -391,6 +535,17 @@ boolean addConnection(C connection) {
                     }
                     if (i == existing.length) {
                         break;
+                    } else if (existing.length == 1 && state == State.EXPIRED) {
```

Comment:
IIUC if we don't observe `state == State.EXPIRED` we may retain a stale entry in our collection until another SD event comes in pertaining to this host?

Reply:
If the state is not `EXPIRED` it is either `CLOSED` or `AVAILABLE`. For `CLOSED` the host is already being removed and its connections are either removed or just about to be removed. For `AVAILABLE` it's not a stale entry – the host can be selected and a new connection will be created upon demand.

Reply:
1. Consider saving `state` in a local variable and break earlier if it's already `CLOSED`.
2. Remove `state == State.EXPIRED` check from here and go inside this branch for all `existing.length == 1` cases. Then try to `cas(EXPIRED, CLOSED)`, if fails try `cas(existing, EMPTY_ARRAY)`. It will help to avoid allocating a new object array if host is active.

Reply:
With the new approach suggested by @Scottmitch this optimisation won't be necessary.